### PR TITLE
Validator

### PR DESCRIPTION
**Pull request by Kalololol**
Create a validator for the "TransacDTO" and "TransacDTOList" classes.
Added validated in templates, controller, created tests.

### CODE REVIEW

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java [hunk 1]
```diff
@@ -4,77 +4,92 @@
...
         this.transacService = transacService;
     }
-
-    //adnotacja aby sprawdzać czy dane nie są puste bądź nullem ---np noEmpty
     @GetMapping("/singleTransaction")
     public String showToSingleTransaction(Model model) {
         try {
             model.addAttribute("transacDTO", new TransacDTO());
             return "singleTransaction";
         }catch (Exception e){
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: Can you make the exception less generic?

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java [hunk 2]
```diff
@@ -4,77 +4,92 @@
...
+        try {
+          TransacDTO transac = transacService.addTransaction(transacDTO);
           model.addAttribute("transac", transac);
           return "singleTransaction";
-      }catch (Exception e){
+      }catch (ConstraintViolationException cve){
+            cve.printStackTrace();
+            return "singleTransaction";
+        }
+        catch (Exception e){
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: Can you make the exception less generic?

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java [hunk 3]
```diff
@@ -4,77 +4,92 @@
...
-        model.addAttribute("newTransacDTO", new TransacDTO());
 
-        return "manyTransactions";
+    @GetMapping("/manyTransactions")
+    public String addToManyTransactions(final Model model){
+        try {
+            model.addAttribute("transacDTOList", transacDTOList);
+            model.addAttribute("newTransacDTO", new TransacDTO());
+            return "manyTransactions";
+        }catch (Exception e){
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: Can you make the exception less generic?

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java [hunk 4]
```diff
@@ -4,77 +4,92 @@
...
-    public String showCreateFormTransactions(@ModelAttribute TransacDTO transacDTO){
+    public String showCreateFormTransactions(final @ModelAttribute(name = "transacDTO") TransacDTO transacDTO, final BindingResult bindingResult){
+        if (bindingResult.hasErrors()) {
+            return "manyTransactions";
+        }
         try {
             transacDTOList.add(transacDTO);
-
             return "redirect:/manyTransactions";
         }catch (Exception e){
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: Can you make the exception less generic?

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java [hunk 5]
```diff
@@ -4,77 +4,92 @@
...
-    public String showMultipleTransactions(Model model) {
+    public String showMultipleTransactions(final Model model) {
         try {
-            List<Transac> resultTransacList = transacService.addManyTransactions(transacDTOList);
+            List<TransacDTO> resultTransacList = transacService.addManyTransactions(transacDTOList);
             transacDTOList.clear();
-
             model.addAttribute("resultTransacList", resultTransacList);
             return "showResultTransactions";
         }catch (Exception e){
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: Can you make the exception less generic?

File: src/main/java/com/example/ForexApp_v1/model/TransacDTOList.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.List;
+
+import com.example.ForexApp_v1.logic.validator.ListTransacDTOConstraint;
+import jakarta.validation.Valid;
+import jakarta.validation.constraints.*;
+
+
+public class TransacDTOList {
+    @Valid
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/model/TransacDTOList.java]: Is this annotation necessary?

File: src/main/java/com/example/ForexApp_v1/service/TransacService.java [hunk 7]
```diff
@@ -27,13 +29,16 @@ public TransacService(TransacRepository transacRepository, CurrencyDownloadApi c
         this.jsonMapper = jsonMapper;
     }
 
-    public Transac addTransaction(TransacDTO transacDTO){
+    public TransacDTO addTransaction(TransacDTO transacDTO){
         Transac transac = calculateTransaction.manualCalculate(transacDTO.getValueCurrency(), transacDTO.getCodeCurrency(), (transacDTO.getDateTransaction().toString()).substring(0,10));
         transacRepository.createOrUpdate(transac);
-        return transac;
+
+        TransacDTO result = new TransacDTO(transac.getDateTransaction().toString(), transac.getCodeCurrency(), transac.getValueCurrency(), transac.getId(), transac.getValuePln(), transac.getResultTransaction(), transac.getIsDone());
```
Comment (blazejprzyluski) [inline src/main/java/com/example/ForexApp_v1/service/TransacService.java]: you could just return the value of new TransacDTO, no need for a variable in the meantime

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import jakarta.validation.ConstraintValidatorContext;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class TransacListValidatorTest {
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java:15]: Please add more test cases, as validator checks multiple things

File: src/main/java/com/example/ForexApp_v1/service/TransacService.java (hunk 7 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/service/TransacService.java]: you're right! thanks :)

File: src/main/java/com/example/ForexApp_v1/model/TransacDTOList.java (hunk 6 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/model/TransacDTOList.java]: this is unnecessary

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java (hunk 1 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: Ok

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java (hunk 2 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: ok

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java (hunk 3 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: ok

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java (hunk 4 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: ok

File: src/main/java/com/example/ForexApp_v1/controller/TransactionController.java (hunk 5 above)
Comment (Kalololol) [reply src/main/java/com/example/ForexApp_v1/controller/TransactionController.java]: ok

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java (hunk 8 above)
Comment (Kalololol) [reply src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java:15]: done

**Comment by Kalololol** [comment]
changes made, if you can, please look at "getCurrencyExchangeRateCompareToWrongDateTest" - is the getCurrencyExchangeRate method correct?

**Comment by blazejprzyluski** [comment]
Is this test part of this PR? I can't find it

**Comment by Kalololol** [comment]
look 
https://github.com/Kalololol/ForexApp_v1/pull/1/commits 
https://github.com/Kalololol/ForexApp_v1/pull/1/commits/6f8f33d35698b11048bbe9b469b2e2f9f3c4ede3

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    @Test
+    public void isValidNullListFalseTest(){
+
+        TransacDTOList transacDTOList = new TransacDTOList();
+        List<TransacDTO> listTransac = null;
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+        TransacDTO transacDTO3 = new TransacDTO("2024-02-21", "", 321.00);
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+        TransacDTO transacDTO3 = new TransacDTO("2024-02-21", null, 232.33);
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+        TransacDTO transacDTO3 = new TransacDTO(null, "USD", 232.33);
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+        TransacDTO transacDTO3 = new TransacDTO("", "USD", 232.33);
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+        TransacDTO transacDTO3 = new TransacDTO("2024-02-21", "USD", null);
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java:84]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+            TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+            TransacDTO transacDTO3 = new TransacDTO("2024-02-21", "USD");
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        TransacDTO transacDTO2 = new TransacDTO("2024-02-21", "USD", 123.00);
+        TransacDTO transacDTO3 = new TransacDTO("2024-02-21", "USD", 88.88);
+        listTransac.add(transacDTO1);
+        listTransac.add(transacDTO2);
+        listTransac.add(transacDTO3);
+        transacDTOList.setTransacDTOList(listTransac);
+
+        ConstraintValidatorContext context = Mockito.mock(ConstraintValidatorContext.class);
+
+        boolean result = transacListValidator.isValid(transacDTOList, context );
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java]: no need for this variable

File: src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
```
Comment (blazejprzyluski) [inline src/test/java/com/example/ForexApp_v1/TransacListValidatorTest.java:33]: no need for this variable

**Comment by Kalololol** [comment]
Done

**Pull request merged**